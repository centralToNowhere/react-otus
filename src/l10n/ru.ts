export const l10n = {
  maxWidthLabel: "Макс. ширина:",
  maxHeightLabel: "Макс. высота:",
  cellSizeLabel: "Размер ячейки:",
  capacityLabel: "Процент заполненности:",
  speedLabel: "Обновлений в секунду:",
  buttonStart: "Старт",
  buttonStop: "Стоп",
  buttonReset: "Сброс",
  buttonGenerateRandom: "Создать",

  // Экран регистрации
  gameNameFull: "Conway's Game of Life",
  gameHeadingPart1: "Conway's",
  gameHeadingPart2: "Game of Life",
  buttonStartGameAsPlayer: "Начать игру",
  rulesHeading: "Правила",
  description1: "Игра «Жизнь» (англ. ",
  source: "Источник:",
  sourceName: "Игра «Жизнь»",
  description2:
    ") — клеточный автомат, придуманный английским математиком Джоном Конвеем в 1970 году.",
  rule1:
    "Место действия этой игры — «вселенная» — это размеченная на клетки поверхность или плоскость — безграничная, ограниченная, или замкнутая (в пределе — бесконечная плоскость).",
  rule2:
    "Каждая клетка на этой поверхности может находиться в двух состояниях: быть «живой» (заполненной) или быть «мёртвой» (пустой). Клетка имеет восемь соседей, окружающих её.",
  rule3:
    "Распределение живых клеток в начале игры называется первым поколением. Каждое следующее поколение рассчитывается на основе предыдущего по таким правилам:",
  rule3_1:
    "в пустой (мёртвой) клетке, рядом с которой ровно три живые клетки, зарождается жизнь",
  rule3_2:
    "если у живой клетки есть две или три живые соседки, то эта клетка продолжает жить; в противном случае, если соседей меньше двух или больше трёх, клетка умирает («от одиночества» или «от перенаселённости»)",
  rule4: "Игра прекращается, если:",
  rule4_1: "на поле не останется ни одной «живой» клетки",
  rule4_2:
    "конфигурация на очередном шаге в точности (без сдвигов и поворотов) повторит себя же на одном из более ранних шагов (складывается периодическая конфигурация)",
  rule4_3:
    "при очередном шаге ни одна из клеток не меняет своего состояния (складывается стабильная конфигурация; предыдущее правило, вырожденное до одного шага назад)",

  // Контейнер игрока
  registerPlayerLabel: "Введите имя игрока",
  logoutButton: "Выйти",
  generation: "Поколение",

  // Ошибки
  minCellsAmount: "Размер игрового поля меньше размера ячейки.",
  maxCellsAmount: "Превышено максимальное количество ячеек",
  positiveNumber: "Ожидается положительное число.",
  nonNegativeNumber: "Ожидается неотрицательное число.",
};
