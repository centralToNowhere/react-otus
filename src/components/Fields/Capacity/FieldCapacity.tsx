import React, { useEffect, useMemo, useState } from "react";
import { l10n } from "@/l10n/ru";
import {
  FieldValidator,
  InputField,
  LabelField,
  onBlurHandler,
  useOnChangeHandler,
} from "@/components/Fields";
import { FormField } from "@/components/Form/FormField";
import { FieldError } from "@/components/Fields/FieldError/FieldError";
import { FormContainer, IFieldProps } from "@/components/Form";
import { useDebounce, isValidNonNegativeNumericString } from "@/utils";
import {
  onRawBlurHandler,
  onRawChangeHandler,
} from "@/components/Fields/FieldHandlers";
import styled from "@emotion/styled";

export const FieldCapacity: React.FC<
  IFieldProps<{
    rawCapacity: string;
  }>
> = (props) => {
  const [error, setError] = useState({
    show: false,
    msg: "",
  });

  const capacityValidator: FieldValidator = useMemo(
    () => ({
      validator: (value: unknown) => isValidNonNegativeNumericString(value),
      setError,
    }),
    []
  );

  const onChangeDebounced = useDebounce<string>(
    useOnChangeHandler(props.onChange, capacityValidator, props.onRawChange),
    FormContainer.inputDelay
  );

  const onChange = onRawChangeHandler((value) => {
    props.onRawChange(value);
    onChangeDebounced(value);
  });

  const onBlur = onRawBlurHandler((value) => {
    onChangeDebounced.clear();
    onBlurHandler(props.onChange, capacityValidator)(value);
  });

  useEffect(() => {
    return () => {
      onChangeDebounced.clear();
    };
  }, [onChangeDebounced]);

  return (
    <FormField>
      <LabelField htmlFor="capacity-percentage">
        {l10n.capacityLabel}
      </LabelField>
      <p>{props.formRawData.rawCapacity} %</p>
      <InputFieldCapacity
        id="capacity-percentage"
        type="range"
        max="100"
        min="0"
        step="1"
        name="capacityPercentage"
        value={props.formRawData.rawCapacity}
        autoComplete="off"
        onChange={onChange}
        onBlur={onBlur}
      />
      <FieldError show={error.show} msg={error.msg} />
    </FormField>
  );
};

const InputFieldCapacity = styled(InputField)`
  border: 0;
  &:focus {
    border-bottom: 0;
  }
`;
