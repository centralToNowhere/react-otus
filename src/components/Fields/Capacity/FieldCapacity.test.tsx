import React from "react";
import { render, screen } from "@testing-library/react";
import { FieldCapacity } from "./FieldCapacity";
import { l10n } from "@/l10n/ru";

describe("FieldCapacity tests", () => {
  it("should render capacity input", () => {
    const { asFragment } = render(
      <FieldCapacity
        formRawData={{
          rawCapacity: "50",
        }}
        onRawChange={() => {
          // empty function
        }}
        onChange={() => {
          // empty function
        }}
      />
    );

    const input: HTMLInputElement = screen.getByLabelText(l10n.capacityLabel);

    expect(input).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
