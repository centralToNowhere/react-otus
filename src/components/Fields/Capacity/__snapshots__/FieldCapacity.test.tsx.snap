// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldCapacity tests should render capacity input 1`] = `
<DocumentFragment>
  <div
    class="css-w6s8l8-FormField e1jhpckf0"
  >
    <label
      class="css-m0dhjx-LabelField e16iqet90"
      for="capacity-percentage"
    >
      Процент заполненности:
    </label>
    <p>
      50 %
    </p>
    <input
      autocomplete="off"
      class="css-1qphvf3-InputField-InputFieldCapacity e1ec6q4r0"
      id="capacity-percentage"
      max="100"
      min="0"
      name="capacityPercentage"
      step="1"
      type="range"
      value="50"
    />
  </div>
</DocumentFragment>
`;
